guides/micronaut-graphql-todo/micronaut-graphql-todo.adoc

```diff
@@ -0,0 +1,398 @@
+
+source:ToDo[]
+
+=== Repositories
+
+To allow fetching of these Entities from the database, we can create https://micronaut-projects.github.io/micronaut-data/latest/guide/#dbcRepositories[`JdbcRepositories`] for each.
+
+The simplest of these is the `ToDoRepository` which just requires the default access methods:
+
+source:ToDoRepository[]
+
+<1> We need to specify the dialect, as queries are computed at compile time.
+
+Then we can specify a Repository for our Authors.  We need a few extra finders for this so that we can hook it into GraphQL more easily in the next step:
+
+source:AuthorRepository[]
+
+<1> Again, we need to specify the dialect.
+<2> When creating todos we need to search for an exisiting username.
+<3> When GraphQL loads a ToDo, we will use this to fetch the authors if they are required in the response.
+<4> We need to find an existing author, or else create a new one when creating a ToDo.
+
+== GraphQL
+
+The initial Micronaut application `create-app` step already added the GraphQL dependency into our build for us:
+
+dependency:micronaut-graphql[groupId=io.micronaut.graphql]
+
+So the default GraphQL endpoint `/graphql` is enabled, and you don't need to add any extra configuration.
+
+=== Describe your schema
+
+Create the file `schema.graphqls` in `src/main/resources` directory:
+
+resource:schema.graphqls[]
+
+<1> Declare a `toDos` function to fetch all the todos as a list.
+<2> Declare a `createToDo` function to create a new ToDo (and return it)
+<3> Declare a `completeToDo` function to mark a ToDo as done (and return a boolean if successful)
+<4> Declare a `ToDo` type
+<5> Declare an `Author` type
+
+=== Data Fetchers
+
+For each of our queries and mutators in the schema, we create a `DataFetcher` which we will bind to the GraphQL schema and our domain model, and execute the appropriate queries in our datastore.
+
+==== Queries
+
+Create class `ToDosDataFetcher` to implement our `toDos` query:
+
+source:ToDosDataFetcher[]
+
+==== Mutations
+
+Create class `CreateToDoDataFetcher` for the creation of ToDos:
+
+source:CreateToDoDataFetcher[]
+
+<1> Find the existing author or create a new one.
+<2> Save the new ToDo in the database.
+
+And a class `CompleteToDoDataFetcher` to mark ToDos as complete:
+
+source:CompleteToDoDataFetcher[]
+
+<1> Find the existing ToDo based on it's id.
+<2> If found, mark it as completed.
+<3> And update the version in the database.
+
+==== Wiring
+
+GraphQL allows us to fetch data on demand.
+In this example, a user may request a list of ToDos, but not require the author to be populated.
+This means we need a way of optionally loading Authors based on their ID.
+
+To do this, we can register a DataLoader that will find authors based on a set of Ids:
+
+source:AuthorDataLoader[]
+
+We can then register this loader in our DataLoaderRegistry under the key `author`
+
+source:DataLoaderRegistryFactory[]
+
+<1> This registry will have request scope, so will be recreated for every request.
```

Comment:
"This registry will have request scope, so will be recreated for every request" -> "This registry has request scope, so a new one will be created for every request"